crates/flux-refineck/src/lib.rs

```diff
@@ -104,8 +104,11 @@ pub fn check_fn(
             .emit(&genv)?;
 
         // PHASE 1: infer shape of `TypeEnv` at the entry of join points
-        let shape_result =
-            Checker::run_in_shape_mode(genv, local_id, &ghost_stmts, config).emit(&genv)?;
+        let shape_result = Checker::run_in_shape_mode(genv, local_id, &ghost_stmts, config)
+            // Augment the possible CheckError with the functions span so we can report
+            // helpful error messages for opaque struct field accesses
+            .map_err(|x| x.with_fn_span(genv.tcx().def_span(def_id)))
```

Comment:
@nilehmann, my idea was to augment `CheckerError` with the functions span, so we can use it within the `Diag` implementation for `OpaqueStruct` errors.

Reply:
I'm not a fan of that because the span can be `None` and you have to reason about it, but in practice we always have it. Instead, we ca do:
```rust
let shape_result = Checker::run_in_shape_mode(genv, local_id, &ghost_stmts, config)
    .map_err(|err| emit_checker_err(genv, err, def_id)))?;
    
fn emit_checker_err(genv: GlobalEnv, err: CheckerError, def_id: DefId) -> ErrorGuaranteed {
   ...
}
``` 
and prove statically that we have a `def_id`.

Alternatively, we can define `emit` as a method in `CheckerError`
```rust
fn emit(self, genv: GlobalEnv, def_id: DefId) -> ErrorGuaranteed;
 ```
 
 You can get a `DiagCtxtHandle` with `genv.sess().dcx().handle()`

Reply:
I don't feel too strongly about this though. I'll let you decide.

Reply:
hmm ya. I agree that it's a bit annoying, but wouldn't we have to basically reimplement the logic in CheckerError's implementation of `Diagnostic`?

Reply:
Yes, I suggest we remove the `Diagnostic` implementation for `CheckerError` and instead have a method/function that takes the def id and emits the error.

Reply:
Ah ok - I think I've got it. The only thing I'm wondering is what the level should be for the `Query` variant of `CheckerErrKind`. My guess is that it should be `Error`? Or is there a way to get that for the `CheckerError`?